File: Solution/1620.网络信号最好的坐标/1620.网络信号最好的坐标.py
'''
Descroption: LeetCode 1620. 网络信号最好的坐标
Author: EmoryHuang
Date: 2022-11-02 09:20:30
Method:
暴力枚举
由于数据范围较小，可以直接暴力枚举每个坐标的信号强度
'''


class Solution:

    def bestCoordinate(self, towers: List[List[int]], radius: int) -> List[int]:

        def t2q(i, j, x, y, q):
            dis = math.sqrt((i - x)**2 + (j - y)**2)
            if dis > radius:
                return 0
            else:
                return q // (1 + dis)

        ans, max_q = [0, 0], 0
        for i in range(51):
            for j in range(51):
                cur_q = sum([t2q(i, j, x, y, q) for x, y, q, in towers])
                if cur_q > max_q:
                    max_q = cur_q
                    ans = [i, j]
        return ans

File: Solution/1620.网络信号最好的坐标/1620.网络信号最好的坐标.cpp
/*
 * @Descroption: LeetCode 1620. 网络信号最好的坐标
 * @Author: EmoryHuang
 * @Date: 2021-11-05 20:11:22
 * @Method:
 * 遍历所有点，找到网络信号最好的坐标
 */

class Solution {
   public:
    vector<int> bestCoordinate(vector<vector<int>>& towers, int radius) {
        vector<int> ans = {0, 0};
        int maxq = 0;
        for (int i = 0; i < 51; i++) {
            for (int j = 0; j < 51; j++) {
                int cur = 0;
                for (auto t : towers) {
                    int x = t[0], y = t[1], q = t[2];
                    double d = sqrt((x - i) * (x - i) + (y - j) * (y - j));
                    if (d <= radius) cur += (int)q / (1 + d);
                }
                if (cur > maxq) {
                    maxq = cur;
                    ans[0] = i;
                    ans[1] = j;
                }
            }
        }
        return ans;
    }
};